src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java

```diff
@@ -0,0 +1,180 @@
+    this.joinDefinition = joinDefinition;
+    this.bigQuery = bigQuery;
+    this.project = project;
+    this.bqDataset = bqDataset;
+    this.bqTable = bqTable;
+    this.jobId = jobId;
+    this.helper = new BigQuerySQLBuilder(this.project, this.bqDataset, stageToTableNameMap);
+  }
+
+  public static BigQueryJoinDataset getInstance(SQLJoinRequest joinRequest,
+                                                Map<String, String> bqTableNamesMap,
+                                                BigQuery bigQuery,
+                                                String project,
+                                                String dataset,
+                                                String runId) {
+
+    // Get new Job ID for this push operation
+    String jobId = BigQuerySQLEngineUtils.getIdentifier();
+
+    // Build new table name for this dataset
+    String table = BigQuerySQLEngineUtils.getNewTableName(runId);
+
+    BigQueryJoinDataset instance = new BigQueryJoinDataset(joinRequest.getDatasetName(),
+                                                           joinRequest.getJoinDefinition(),
+                                                           bqTableNamesMap,
+                                                           bigQuery,
+                                                           project,
+                                                           dataset,
+                                                           table,
+                                                           jobId);
+    instance.executeJoin();
+    return instance;
+  }
+
+  private void executeJoin() {
+    TableId destinationTable = TableId.of(project, bqDataset, bqTable);
+
+    String query = getQuery(joinDefinition);
+    LOG.info("Creating table `{}` using job: {} with SQL statement: {}", bqTable, jobId, query);
+
+    // Run BigQuery job with generated SQL statement, store results in a new table, and set priority to BATCH
+    QueryJobConfiguration queryConfig =
+      QueryJobConfiguration.newBuilder(query)
+        .setDestinationTable(destinationTable)
+        .setPriority(QueryJobConfiguration.Priority.BATCH)
+        .build();
+
+    // Create a job ID so that we can safely retry.
+    JobId bqJobId = JobId.of(jobId);
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(bqJobId).build());
+
+    // Wait for the query to complete.
+    try {
+      queryJob = queryJob.waitFor();
+    } catch (InterruptedException ie) {
+      throw new SQLEngineException("Interrupted exception when executing Join operation", ie);
+    }
+
+    // Check for errors
+    if (queryJob == null) {
+      throw new SQLEngineException("BigQuery job not found: " + jobId);
+    } else if (queryJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      throw new SQLEngineException("Error executing BigQuery Job: "
```

Comment:
should also include all info (jobid, project?, location?) needed to allow the user to find the actual BQ job to take a closer look.

Reply:
Will add better logging here.